### ISSUE TITLE
Performance Issue with Hierarchical Structure

### ISSUE THREAD

**Issue by ronag** [rackt/redux#764]
I've gotten a bit stuck in porting our application from a Relay like structure to Redux.

Specfically when having a data structure with a selector for `mapStateToProps` like this:

```javascript
export function selectItem(items, id) {
  const item = items[id]
  return {
    id,
    ...item && {
      children: item.children.map((id) => selectItem(items, id))
    }
  }
}

@connect((state) => selectItem(state.items, state.currentItem))
```

In Relay this wouldn't be a problem since every item would have it's own isolated selector. However, the way I am trying to do it now with Redux it becomes a bit of a problem since an update to any single item will cause the entire tree to be rebuilt. 

I'm hoping for a memorisation based solution from here https://github.com/faassen/reselect/issues/47.

However, it feels like I am missing or doing something wrong. Any advice?

**Comment by ronag** [comment]
I have extend on the example here https://github.com/rackt/redux/issues/765

**Comment by gaearon** [comment]
Any reason you closed the issue? Have you found a satisfactory solution?

**Comment by ronag** [comment]
No I did not. I've created a server side service which transforms the data prior to sending to the client. Similar to Falcor and GraphQL.

**Comment by ronag** [comment]
Re-opening this issue. Would be nice if we could figure out how to handle these kind of select transformations in an efficient way.

**Comment by gaearon** [comment]
cc @ellbee @faassen

**Comment by ronag** [comment]
@gaearon: Since you are the author of both `normalizr`, `redux` and `redux-react` I would be interested on your take on the problem over at https://github.com/faassen/reselect/issues/47, specially the last comments.

**Comment by ellbee** [comment]
Resolved in #815
